node/libs/concurrency/src/sync/prunable_mpsc/mod.rs

```diff
@@ -51,17 +67,65 @@ struct Shared<T> {
 #[allow(clippy::type_complexity)]
 pub struct Sender<T> {
     shared: Arc<Shared<T>>,
-    pruning_predicate: Box<dyn Sync + Send + Fn(&T, &T) -> bool>,
+    filter_function: Box<dyn Sync + Send + Fn(&T) -> SelectionFunctionResult>,
+    selection_function: Box<dyn Sync + Send + Fn(&T, &T) -> SelectionFunctionResult>,
 }
 
 impl<T> Sender<T> {
     /// Sends a value.
-    /// This initiates the pruning procedure which operates in O(N) time complexity
+    /// This initiates the filter and pruning procedure which operates in O(N) time complexity
     /// on the buffer of pending values.
     pub fn send(&self, value: T) {
+        // Check if new value is valid to keep
+        if (self.filter_function)(&value) != SelectionFunctionResult::Keep {
+            return;
+        }
         self.shared.send.send_modify(|buf| {
-            buf.retain(|pending_value| !(self.pruning_predicate)(pending_value, &value));
-            buf.push_back(value);
+            let len = buf.len();
+            let mut idx = 0;
+            let mut cur = 0;
+
+            // Stage 1: All values are retained.
+            while cur < len {
+                match (self.selection_function)(&mut buf[cur], &value) {
+                    SelectionFunctionResult::Keep => {
+                        cur += 1;
+                        idx += 1;
+                    }
+                    SelectionFunctionResult::DiscardOld => {
+                        cur += 1;
+                        break;
+                    }
+                    SelectionFunctionResult::DiscardNew => return,
+                }
+            }
+            // Stage 2: Swap retained value into current idx.
+            while cur < len {
+                match (self.selection_function)(&mut buf[cur], &value) {
+                    SelectionFunctionResult::Keep => {
+                        buf.swap(idx, cur);
```

Comment:
https://doc.rust-lang.org/std/collections/struct.VecDeque.html#method.swap idx and cur can be equal. Duplicating the loop is not needed.

Reply:
I'm not sure I follow. If we want this to be efficient, it would be better to have the queue group messages by sender, so that you can lookup them directly when adding a new message, instead of iterating over the whole thing. If we do O(n) insertion, then simplest implementation is a way to go.

Reply:
Yes, it's just a nitpick (to avoid 'same index' swaps). If we group by sender we wouldn't have a generic data structure. I just adapted the [VeqDeque](https://github.com/rust-lang/rust/blob/master/library/alloc/src/collections/vec_deque/mod.rs#L2057-L2089) implementation to have at least same performance as before. 
But already changed the implementation to your suggestion.